components/server/src/user/gitpod-token-service.ts

```diff
@@ -0,0 +1,110 @@
+import { log } from "@gitpod/gitpod-protocol/lib/util/logging";
+import { inject, injectable } from "inversify";
+import { Authorizer } from "../authorization/authorizer";
+import { GuardedResource, ResourceAccessGuard, ResourceAccessOp } from "../auth/resource-access";
+import { ApplicationError, ErrorCodes } from "@gitpod/gitpod-protocol/lib/messaging/error";
+
+@injectable()
+export class GitpodTokenService {
+    constructor(
+        @inject(UserDB) private readonly userDB: UserDB,
+        @inject(Authorizer) private readonly auth: Authorizer,
+    ) {}
+
+    async getGitpodTokens(
+        requestorId: string,
+        userId: string,
+        resouceGuardAccess: ResourceAccessGuard,
+    ): Promise<GitpodToken[]> {
+        await this.auth.checkPermissionOnUser(requestorId, "read_tokens", userId);
+        const res = (await this.userDB.findAllGitpodTokensOfUser(userId)).filter((v) => !v.deleted);
+        await Promise.all(
+            res.map((tkn) => this.guardAccess(resouceGuardAccess, { kind: "gitpodToken", subject: tkn }, "get")),
+        );
+        return res;
+    }
+
+    async generateNewGitpodToken(
+        requestorId: string,
+        userId: string,
+        options: { name?: string; type: GitpodTokenType; scopes?: string[] },
+        resouceGuardAccess: ResourceAccessGuard,
+    ): Promise<string> {
+        await this.auth.checkPermissionOnUser(requestorId, "write_tokens", userId);
+        const token = crypto.randomBytes(30).toString("hex");
+        const tokenHash = crypto.createHash("sha256").update(token, "utf8").digest("hex");
+        const dbToken: DBGitpodToken = {
+            tokenHash,
+            name: options.name,
+            type: options.type,
+            userId,
+            scopes: options.scopes || [],
+            created: new Date().toISOString(),
+        };
+        await this.guardAccess(resouceGuardAccess, { kind: "gitpodToken", subject: dbToken }, "create");
+        await this.userDB.storeGitpodToken(dbToken);
+        return token;
+    }
+
+    async getGitpodTokenScopes(
+        requestorId: string,
+        userId: string,
+        tokenHash: string,
+        resouceGuardAccess: ResourceAccessGuard,
+    ): Promise<string[]> {
+        await this.auth.checkPermissionOnUser(requestorId, "read_tokens", userId);
+        let token: GitpodToken | undefined;
+        try {
+            token = await this.userDB.findGitpodTokensOfUser(userId, tokenHash);
+        } catch (error) {
+            log.error({ userId }, "failed to resolve gitpod token: ", error);
+            return [];
+        }
+        if (!token || token.deleted) {
+            return [];
+        }
+        await this.guardAccess(resouceGuardAccess, { kind: "gitpodToken", subject: token }, "get");
+        return token.scopes;
+    }
+
+    async deleteGitpodToken(
+        requestorId: string,
+        userId: string,
+        tokenHash: string,
+        resouceGuardAccess: ResourceAccessGuard,
```

Comment:
we don't want to pass the guardAccess done here. It should stay in the gitpod-server-impl